### PR TITLE
Support ATen GPU pointwise apply and torch.where.

### PR DESCRIPTION
**Pull request by gchanan**
Like the CPU version, this implements an apply template that is almost identical to the
apply template already in THC, but using the ATen API.  Much of this involves stripping out
the TensorUtils code (which is basically templated ATen-style), although a couple of functions
remain that are apply specific (and thus don't seem worth porting to ATen), namely
overlappingIndices, canUse32BitIndexMath, and getTensorInfo.  We can make those generally
available via native functions if there's a need.

### CODE REVIEW

**Comment by gchanan** [comment]
A couple of notes/questions for @ngimel:

1) The CPU apply code uses lambdas; to get the CUDA equivalent device lambdas I enabled `--expt-extended-lambda`; is that something we want to do?  Or should we just write fully-fledged functions?

2) NVCC can't seem to handle the previous dispatch templates (you can see them in Dispatch.h), so I had to make the return type an explicit template parameter rather than `auto`.  I spent some time coming up with a minimal reproducible example and it seems like it's the combination of variadic template parameters, non-variadic template parameters, and `auto` return type.  Specifically, something like this seems to work:

```
template <typename T>
struct Foobar {
  static int64_t apply() {return 0;}
};

template<typename ... Args>
auto dispatch_all(Args&&... args)
  -> decltype(Foobar<double>::apply(std::forward<Args>(args)...)) {
...
}

dispatch_all();
```

but adding an unused template parameter (F) doesn't:
```
template <typename T>
struct Foobar {
  static int64_t apply() {return 0;}
};

template<typename F, typename ... Args>
auto dispatch_all(const Type& the_type, const char *name, Args&&... args)
  -> decltype(Foobar<double>::apply(std::forward<Args>(args)...)) {

struct Foo {};
dispatch_all<Foo>();
```

Here's my nvcc version:
```nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2016 NVIDIA Corporation
Built on Sun_Sep__4_22:14:01_CDT_2016
Cuda compilation tools, release 8.0, V8.0.44```

**Review by colesbury — APPROVED** [review]
looks good

**Comment by ngimel** [comment]
@gchanan, 
1) lambda support is still experimental, as that compiler switch says, but it works for the most part. File bugs if something does not work
2) The following snippet (minimally modified from yours to make it compile with g++)
```
#include <stdint.h>
#include <utility>

template <typename T>
struct Foobar {
  static int64_t apply() {return 0;}
};


template<typename ... Args>
auto dispatch_all(Args&&... args)
  -> decltype(Foobar<double>::apply(std::forward<Args>(args)...)) {
  return 0;
}

template<typename F, typename T, typename ... Args>
auto dispatch_all_F(const T& the_type, const char *name, Args&&... args)
  -> decltype(Foobar<double>::apply(std::forward<Args>(args)...)) {
  return 0;

}

struct Foo{};

int main(){
dispatch_all();
float a;
const char * name;
dispatch_all_F<Foo, float>(a,name);

}
```
compiles for me both with g++ and nvcc (9.0). Let me know if I'm doing something wrong here or misunderstanding the patterns you need.

File: aten/src/ATen/native/cuda/NativeFunctionsCuda.cu [hunk 1]
```diff
@@ -1,9 +1,30 @@
...
 #include <cfloat>
 
 namespace at {
 namespace native {
 
+template <typename scalar>
+struct WhereOpCUDA {
+  static void apply(Tensor& ret, const Tensor& condition, const Tensor& self, const Tensor& other) {
+    // yes this name is repetitive, but the CPU version is called CPU_tensor_apply4 and we don't have
+    // a CPU namespace or diectory.
```
[the remark this answers is not in the document]
Comment (gchanan) [reply aten/src/ATen/native/cuda/NativeFunctionsCuda.cu]: I'm not against it, but there are other things that should probably go there (CPUFixedAllocator, CPUGenerator), so let's defer it from this PR.

File: aten/src/ATen/cuda/detail/IndexUtils.cuh [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+#include "ATen/ATen.h"
+#include "TensorInfo.cuh"
+
+namespace at {
+namespace cuda {
+namespace detail {
+
+bool overlappingIndices(const at::Tensor& t);
+bool canUse32BitIndexMath(const at::Tensor &t, ptrdiff_t max_elem=UINT32_MAX);
```
[the remark this answers is not in the document]
Comment (gchanan) [reply aten/src/ATen/cuda/detail/IndexUtils.cuh]: good idea.

File: aten/src/ATen/Dispatch.h [hunk 3]
```diff
@@ -4,42 +4,44 @@
...
-    }
+
+template<typename R, template <typename> class F, typename ... Args>
+R dispatch_floating_types(const Type& the_type, const char *name, Args&&... args) {
+  switch(the_type.scalarType()) {
+    case ScalarType::Double:
+      return F<double>::apply(std::forward<Args>(args)...);
+    case ScalarType::Float:
+      return F<float>::apply(std::forward<Args>(args)...);
+    case ScalarType::Half: // no native half math on either CPU or GPU.
```
[the remark this answers is not in the document]
Comment (gchanan) [reply aten/src/ATen/Dispatch.h]: why? I'm just trying to be explicit that Half isn't dispatched.  It's essentially a comment.

**Comment by gchanan** [comment]
@ngimel it does indeed work on cuda 9.0 (thanks!), but not on cuda 8.0 (I didn't try cuda 7.5).  Since we still support those older versions, we can't use the feature for now.

**Pull request merged**